File: backend/src/modules/auth/auth.service.ts
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { ConflictException, Injectable, InternalServerErrorException, UnauthorizedException } from '@nestjs/common';
import { PrismaService } from 'src/prisma/prisma.service';
import { RegisterDto } from './dto/register.dto';
import { AuthResponseDto } from './dto/auth-response.dto';
import * as bcrypt from 'bcrypt';
import {JwtService} from '@nestjs/jwt';
import { randomBytes } from 'crypto';
import { LoginDto } from './dto/login.dto';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class AuthService {
  private readonly SALT_ROUNDS = 12;

  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService, 
    private configService: ConfigService,
    ) {}

  //register a new user
  async register(registerDto: RegisterDto): Promise<AuthResponseDto> {
    const { email, password, firstName, lastName } = registerDto;

    const existingUser = await this.prisma.user.findUnique({
      where: { email },
    });
    if (existingUser) {
      throw new ConflictException('User with this email already exists');
    }
    try {
      const hashedPassword = await bcrypt.hash(password, this.SALT_ROUNDS);
      const user = await this.prisma.user.create({
        data: {
            email, 
            password: hashedPassword,
            firstName,
            lastName
        },
        select: {
            id: true,
            email: true,
            firstName: true,
            lastName: true,
            role: true,
            password: false
        }
      })

      const tokens = await this.generateTokens(user.id, user.email);
      await this.updateRefreshToken(user.id, tokens.refreshToken);

      return {
        ...tokens,
        user,
      }
    } catch (error) {  
      console.error("Error during user registration: ", error);
      throw new InternalServerErrorException('Registratoin failed')
    }
  }

  //Generate access and refresh tokens
  private async generateTokens(userId: string, email: string): Promise<{accessToken: string; refreshToken: string}>{
    // Implementation for token generation
    const payload = { sub: userId, email}
    const refreshId = randomBytes(16).toString('hex'); 
    const [accessToken, refreshToken] = await Promise.all([
        this.jwtService.signAsync(payload, {expiresIn: '15m'}),
        this.jwtService.signAsync({...payload, refreshId}, {
            secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
            expiresIn:'7d'
        }),
    ])  
     
    return { accessToken, refreshToken };
  }
  
  async updateRefreshToken(userId: string, refreshToken: string): Promise<void>{
    const hash = await bcrypt.hash(refreshToken, this.SALT_ROUNDS);
    await this.prisma.user.update({
        where: { id: userId },
        data: { refreshToken: hash }
    });
  }

  //Refresh access token
  async refreshTokens(userId : string) : Promise<AuthResponseDto>{
    const user = await this.prisma.user.findUnique({
      where: { id : userId},
      select:{
        id: true,
        email: true,
        firstName: true,
        lastName: true,
        role: true  
      }
    })
    if(!user){
      throw new UnauthorizedException("User Not found");
    }

    const tokens = await this.generateTokens(user.id, user.email);
    await this.updateRefreshToken(user.id, tokens.refreshToken);

    return {
      ...tokens,
      user,
    }
  }

  //logout user and invalidate refresh token
  async logout(userId: string) : Promise<void>{
    await this.prisma.user.update({
      where: {id: userId},
      data: {refreshToken: null},
    })
  }

  //login
  async login(loginDto: LoginDto): Promise<AuthResponseDto>{
    const { email, password } = loginDto;
     
    const user = await this.prisma.user.findUnique({
      where: { email },
    })

    if(!user || !(await bcrypt.compare(password, user.password))){
      throw new UnauthorizedException('Invalid email or password');
    }

    const tokens = await this.generateTokens(user.id, user.email)
    await this.updateRefreshToken(user.id, tokens.refreshToken)

    return{
      ...tokens,
      user: {
        id: user.id,
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
        role: user.role,
      }
    }
  }
}

File: ecommerce/src/main/java/com/ecommerce/service/AuthService.java
package com.ecommerce.service;

import com.ecommerce.dto.request.LoginRequest;
import com.ecommerce.dto.request.RegisterRequest;
import com.ecommerce.dto.response.AuthResponse;
import com.ecommerce.entity.User;
import com.ecommerce.entity.enums.Role;
import com.ecommerce.exception.EmailAlreadyExistsException;
import com.ecommerce.repository.UserRepository;
import com.ecommerce.security.jwt.JwtService;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

/**
 * Mirrors NestJS AuthService exactly:
 *  - register()       → hash password (bcrypt, 12 rounds), generate tokens, hash+store refresh token
 *  - login()          → bcrypt.compare password, generate tokens, hash+store refresh token
 *  - refreshTokens()  → called after RefreshTokenFilter has already validated via bcrypt.compare
 *  - logout()         → set refreshToken = null in DB
 *
 * Refresh token rotation: every login/register/refresh generates a NEW refresh token
 * and overwrites the stored bcrypt hash — old refresh tokens are immediately invalidated.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final UserRepository userRepository;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;

    // ── Register ──────────────────────────────────────────────────────────────

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new EmailAlreadyExistsException("User with this email already exists");
        }

        User user = User.builder()
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword())) // bcrypt, 12 rounds
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .role(Role.USER)
                .build();

        userRepository.save(user);
        log.debug("Registered new user: {}", user.getEmail());

        return issueTokens(user);
    }

    // ── Login ─────────────────────────────────────────────────────────────────

    @Transactional
    public AuthResponse login(LoginRequest request) {
        // AuthenticationManager calls bcrypt.compare internally via DaoAuthenticationProvider
        Authentication auth = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        User user = (User) auth.getPrincipal();
        log.debug("User logged in: {}", user.getEmail());

        return issueTokens(user);
    }

    // ── Refresh tokens ────────────────────────────────────────────────────────

    /**
     * Called after RefreshTokenFilter has already:
     *  1. Parsed the token with JWT_REFRESH_SECRET
     *  2. Done bcrypt.compare(rawToken, storedHash)
     * So here we just rotate — generate new pair and hash+store the new refresh token.
     */
    @Transactional
    public AuthResponse refreshTokens(User user) {
        log.debug("Rotating tokens for user: {}", user.getEmail());
        return issueTokens(user);
    }

    // ── Logout ────────────────────────────────────────────────────────────────

    @Transactional
    public void logout(User user) {
        // Set refreshToken = null, matching: prisma.user.update({ data: { refreshToken: null } })
        userRepository.clearRefreshToken(user.getId());
        log.debug("Cleared refresh token for user: {}", user.getEmail());
    }

    // ── Token issuance & rotation ─────────────────────────────────────────────

    /**
     * Generates access + refresh tokens, then bcrypt-hashes and stores the refresh token.
     * Mirrors NestJS:
     *   const tokens = await this.generateTokens(...)
     *   await this.updateRefreshToken(userId, tokens.refreshToken)  // hashed inside updateRefreshToken
     */
    private AuthResponse issueTokens(User user) {
        String accessToken = jwtService.generateAccessToken(user);
        String rawRefreshToken = jwtService.generateRefreshToken(user);

        // Hash the raw refresh token before persisting — mirrors bcrypt.hash in NestJS
        String hashedRefreshToken = passwordEncoder.encode(rawRefreshToken);
        userRepository.updateRefreshToken(user.getId(), hashedRefreshToken);

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(rawRefreshToken)   // raw token returned to client
                .userId(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .role(user.getRole())
                .build();
    }
}
